packages/docusaurus-theme-common/src/hooks/useCodeWordWrap.ts

```diff
@@ -47,10 +59,28 @@ export function useCodeWordWrap(): {
       passive: true,
     });
 
+    if (!mutationObserver) {
+      setMutationObserver(
+        new MutationObserver((mutations) => {
+          mutations.forEach((mutation) => {
+            if (
+              mutation.type === 'attributes' &&
+              mutation.attributeName === 'hidden'
+            ) {
+              updateCodeIsScrollable();
+            }
+          });
+        }),
+      );
+    }
+
```

Comment:
it's a good practice in React to use one separate effect per concern (resize, observe mutations...) instead of trying to fit everything into a single effect

Also I suggest to create a reusable abstraction here to encapsulate the IO complexity.

For example: `useMutationObserver()` could be added to theme-common as it would likely be useful in other places: https://www.30secondsofcode.org/react/s/use-mutation-observer

note: we may want to take care of memorizing the options (or making it a static constant) and callback to avoid IO recreation

And you can still extract a custom hook dedicated to this specific case. 

In the end the component code could look as simple as: `useHiddenAttributeMutationObserver(() => updateCodeIsScrollable());`

Reply:
Thank you for the help! I started implementing `useMutationObserver()` but had trouble passing in the target element, since it is constantly changing, and the mutation observer must call observe on the bew element each time it is changed. I followed the [link](https://www.30secondsofcode.org/react/s/use-mutation-observer), but instead of passing in a RefObject<Element> I pass in an an Element instead of a in order to trigger useEffect(). Is this fine?

I also have another (not finalized, but functioning) implementation that does not need to store any additional state, which I pasted below. It does cause errors with the linter (@typescript-eslint/no-use-before-define), but it works fine. Is it better to go with this method?

```typescript
import type {RefObject} from 'react';
import {useState, useCallback, useEffect, useRef} from 'react';

const useMutationObserver = (
  callback: (mutations: MutationRecord[]) => void,
  options = {
    attributes: true,
    characterData: false,
    childList: false,
    subtree: false,
  }
) => {
  const mutationObserver = useRef<MutationObserver>
(new MutationObserver(callback));

  return {
    setAncestor: (ancestor: Element | null | undefined): void => {
      if (ancestor) {
        mutationObserver.current.observe(ancestor, options);
      }
    }
  }
};

export function useCodeWordWrap(): {
  readonly codeBlockRef: RefObject<HTMLPreElement>;
  readonly isEnabled: boolean;
  readonly isCodeScrollable: boolean;
  readonly toggle: () => void;
} {
  const [isEnabled, setIsEnabled] = useState(false);
  const [isCodeScrollable, setIsCodeScrollable] = useState<boolean>(false);
  const codeBlockRef = useRef<HTMLPreElement>(null);
  const observer = useMutationObserver((mutations) => {
    mutations.forEach((mutation) => {
      if (
        mutation.type === 'attributes' &&
        mutation.attributeName === 'hidden'
      ) {
        updateCodeIsScrollable();
      }
    });
  });

  const toggle = useCallback(() => {
    const codeElement = codeBlockRef.current!.querySelector('code')!;

    if (isEnabled) {
      codeElement.removeAttribute('style');
    } else {
      codeElement.style.whiteSpace = 'pre-wrap';
      codeElement.style.overflowWrap = 'anywhere';
    }

    setIsEnabled((value) => !value);
  }, [codeBlockRef, isEnabled]);

  const updateCodeIsScrollable = useCallback(() => {
    const {scrollWidth, clientWidth} = codeBlockRef.current!;
    // Allows code block to update scrollWidth and clientWidth after "hidden"
    // attribute is removed
    observer.setAncestor(codeBlockRef.current?.closest('[hidden]'));
    const isScrollable =
      scrollWidth > clientWidth ||
      codeBlockRef.current!.querySelector('code')!.hasAttribute('style');
    setIsCodeScrollable(isScrollable);
  }, [codeBlockRef, observer]);

  useEffect(() => {
    updateCodeIsScrollable();
  }, [isEnabled, updateCodeIsScrollable]);

  useEffect(() => {
    window.addEventListener('resize', updateCodeIsScrollable, {
      passive: true,
    });

    return () => {
      window.removeEventListener('resize', updateCodeIsScrollable);
    };
  }, [updateCodeIsScrollable]);

  return {codeBlockRef, isEnabled, isCodeScrollable, toggle};
}

```